java/ql/src/experimental/Security/CWE/CWE-625/PermissiveDotRegex.ql

```diff
@@ -0,0 +1,40 @@
+/**
+ * @name URL matched by permissive `.` in the regular expression
+ * @description URL validated with permissive `.` in regex  are possibly vulnerable
+ *              to an authorization bypass.
+ * @kind path-problem
+ * @problem.severity warning
+ * @precision high
+ * @id java/permissive-dot-regex
+ * @tags security
+ *       external/cwe-625
+ *       external/cwe-863
+ */
+
+import java
+import semmle.code.java.dataflow.FlowSources
+import DataFlow::PathGraph
+import PermissiveDotRegexQuery
+
+from DataFlow::PathNode source, DataFlow::PathNode sink, MatchRegexConfiguration conf
+where
+  conf.hasFlowPath(source, sink) and
+  exists(MethodAccess ma | any(PermissiveDotRegexConfig conf2).hasFlowToExpr(ma.getArgument(0)) |
+    // input.matches(regexPattern)
+    ma.getMethod() instanceof StringMatchMethod and
+    ma.getQualifier() = sink.getNode().asExpr()
+    or
+    // p = Pattern.compile(regexPattern); p.matcher(input)
+    ma.getMethod() instanceof PatternCompileMethod and
+    exists(MethodAccess pma |
+      pma.getMethod() instanceof PatternMatcherMethod and
+      sink.getNode().asExpr() = pma.getArgument(0) and
+      DataFlow::localExprFlow(ma, pma.getQualifier())
+    )
+    or
+    // p = Pattern.matches(regexPattern, input)
+    ma.getMethod() instanceof PatternMatchMethod and
+    sink.getNode().asExpr() = ma.getArgument(1)
+  )
```

Comment:
I think we could move this code to `MatchRegexConfiguration::isSink` (or an auxiliary predicate called from `MatchRegexConfiguration::isSink`). That should improve performance a bit.

Reply:
Thank @atorralba. I've moved pattern check to `MatchRegexConfiguration::isSink`.

The reason that the Apache Shiro result is no longer shown is:
- 1. To reduce FPs, we changed the matching algorithm to only detect strings ending with `.*` 

from
```
    exists(string s, int i | this.getValue() = s |
      s.indexOf(".") = i and
      not s.charAt(i - 1) = "\\"
    )
```
to
```
    exists(string s, int i | this.getValue() = s |
      s.indexOf(".*") = i and
      not s.charAt(i - 1) = "\\" and
      s.length() = i + 2
    )
```
And the Shiro library only has a pattern example `/index.html` in its test cases e.g. `org.apache.shiro.web.filter.mgt.PathMatchingFilterChainResolverTest`
```
        resolver.getFilterChainManager().addToChain("/index.html", "authcBasic");
```

which is then checked in the vulnerable program `org.apache.shiro.util.RegExPatternMatcher`:
```
    public boolean matches(String pattern, String source) {
       ...
        Pattern p = Pattern.compile(pattern);
        Matcher m = p.matcher(source);
        return m.matches();
    }
```

-  2. To reduce FPs, we added a constraint that the pattern match is used in a forward/redirect context. 
```
  override predicate isSink(DataFlow::Node sink) {
    ....
    exists(
      Guard guard, Expr se, Expr ce // used in a condition to control url redirect, which is a typical security enforcement
      (
```
Since Shiro is an easy-to-use Java security framework that performs authentication, authorization, cryptography, and session management, which is to be integrated in custom applications, we don't have redirect calls in this library.

As such, I think it's probably OK that no query result is returned for this library. Please advise.

Thanks,
@luchua-bc

Reply:
Ok, makes sense, since these were suggestions by the security lab to reduce the FP ratio. Thanks for clarifying.